File: 86-partition-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func partition(head *ListNode, x int) *ListNode {
    if head == nil{
        return head
    }
    var bh, bc, sh, sc *ListNode
    nd := head
    for nd != nil{
        if nd.Val < x{
            if sh == nil{
                sh = nd
                sc = nd
            } else {
                sc.Next = nd
                sc = nd
            }
        } else {
            if bh == nil{
                bh = nd
                bc = nd
            } else {
                bc.Next = nd
                bc = nd
            }
        }
        nd = nd.Next
    }
    if sc != nil {
        sc.Next = bh
    }
    if bc != nil {
        bc.Next = nil
    }
    if sh != nil{
        return sh
    }
    return bh
}

File: bwu/linked_list/86-partition-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def partition(self, head, x):
        """
        :type head: ListNode
        :type x: int
        :rtype: ListNode
        """
        if head is None:
            return head
            
        dummy = ListNode(0)
        dummy.next = head
        dummy2 = ListNode(0)
        lp = dummy2
        prev, curr = dummy, head
        
        while curr:
            if curr.val < x:
                lp.next = curr
                lp = lp.next
                prev.next = curr.next
                curr = prev.next
            else:
                prev = curr
                curr = curr.next
                
        lp.next = dummy.next
        return dummy2.next
